src/rust/multi.comick/src/parser.rs

```diff
@@ -0,0 +1,329 @@
+use aidoku::{
+	error::Result,
+	prelude::format,
+	std::net::HttpMethod,
+	std::net::Request,
+	std::Vec,
+	std::String,
+	Chapter, DeepLink, Filter, FilterType, Manga, MangaContentRating, MangaPageResult, MangaStatus,
+	MangaViewer, Page,
+};
+use alloc::string::ToString;
+extern crate alloc;
+use crate::helper::*;
+
+pub fn parse_manga_list(
+	api_url: String,
+	filters: Vec<Filter>,
+	page: i32,
+) -> Result<MangaPageResult> {
+	let mut included_tags: Vec<String> = Vec::new();
+	let mut excluded_tags: Vec<String> = Vec::new();
+	let mut demographic_tags: Vec<String> = Vec::new();
+	let mut sort_by: String = String::new();
+	let mut title: String = String::new();
+	let mut manga_type: String = String::new();
+	let mut completed: String = String::new();
+	let sort_options = [
+		"",
+		"view",
+		"uploaded",
+		"rating",
+		"follow",
+		"user_follow_count",
+	];
+	let type_options = ["", "jp", "kr", "cn"];
+	let completed_options = ["", "true", "false"];
+	for filter in filters {
+		match filter.kind {
+			FilterType::Title => {
+				title = filter.value.as_string()?.read();
+			}
+			FilterType::Genre => match filter.value.as_int().unwrap_or(-1) {
+				0 => excluded_tags.push(filter.object.get("id").as_string()?.read()),
+				1 => included_tags.push(filter.object.get("id").as_string()?.read()),
+				_ => continue,
+			},
+			FilterType::Select => {
+				let index = filter.value.as_int().unwrap_or(-1) as usize;
+				match filter.name.as_str() {
+					"Sort" => sort_by = String::from(sort_options[index]),
+					"Type" => manga_type = String::from(type_options[index]),
+					"Completed" => completed = String::from(completed_options[index]),
+					_ => continue,
+				}
+			}
+			FilterType::Check => {
+				if filter.value.as_int().unwrap_or(-1) <= 0 {
+					continue;
+				}
+				match filter.name.as_str() {
+					"Shounen" => demographic_tags.push(String::from("1")),
+					"Shoujo" => demographic_tags.push(String::from("2")),
+					"Seinen" => demographic_tags.push(String::from("3")),
+					"Josei" => demographic_tags.push(String::from("4")),
+					_ => continue,
+				}
+			}
+			_ => continue,
+		};
+	}
+
+	if title.is_empty()
+		&& demographic_tags.is_empty()
+		&& included_tags.is_empty()
+		&& excluded_tags.is_empty()
+		&& sort_by.is_empty()
+		&& manga_type.is_empty()
+		&& completed.is_empty()
+	{
+		parse_manga_listing(api_url, String::from("Hot"), page)
+	} else {
+		let url = get_search_url(
+			api_url,
+			title,
+			included_tags,
+			excluded_tags,
+			demographic_tags,
+			manga_type,
+			sort_by,
+			completed,
+			page,
+		);
+		let mut mangas: Vec<Manga> = Vec::new();
+		let json = Request::new(&url, HttpMethod::Get).json().as_array()?;
+		for data in json {
+			if let Ok(data_obj) = data.as_object() {
+				let title = data_obj.get("title").as_string()?.read();
+				let id = data_obj.get("slug").as_string()?.read();
+				let cover = data_obj.get("cover_url").as_string()?.read();
```

Comment:
you should `continue` if any of these don't exist, rather than using `?` and returning.